runtime/lua/vim/lsp/buf.lua

```diff
@@ -249,13 +249,34 @@ end
 ---@param new_name (string) If not provided, the user will be prompted for a new
 ---name using |input()|.
 function M.rename(new_name)
-  -- TODO(ashkan) use prepareRename
-  -- * result: [`Range`](#range) \| `{ range: Range, placeholder: string }` \| `null` describing the range of the string to rename and optionally a placeholder text of the string content to be renamed. If `null` is returned then it is deemed that a 'textDocument/rename' request is not valid at the given position.
   local params = util.make_position_params()
-  new_name = new_name or npcall(vfn.input, "New Name: ", vfn.expand('<cword>'))
-  if not (new_name and #new_name > 0) then return end
-  params.newName = new_name
-  request('textDocument/rename', params)
+  function prepare_rename(err, _, result)
+    if err == nil and result == nil then
+      print('nothing to rename')
```

Comment:
todo/future: i would expect handlers in our lsp client to return a result and if a message needs to be shown, that would be handled by a central dispatcher or something. I see that print() is used in a few other places, e.g. `add_workspace_folder`. Hmm....

Reply:
Just to clarify, `add_workspace_folder` is not a handler but a function used to send a request to a language server